enhancements/machine-api/control-plane.md

```diff
@@ -0,0 +1,281 @@
+
+All of these components are expected to keep self managing themselves as the cluster shrink and expand the Control Plane compute resources.
+
+### User Stories [optional]
+
+#### Story 1
+- As an operator running Installer Provider Infrastructure (IPI), I want flexibility to run [large or small clusters](https://kubernetes.io/docs/setup/best-practices/cluster-large/#size-of-master-and-master-components) so I need to have the ability to resize the control plane in a declarative, automated and seamless manner.
+
+#### Story 2
+- As an operator running User Provider Infrastructure (UPI), I want to expose my non-machine API machines and offer them to the Control Plane controller so I can have the ability to resize the control plane in a declarative, automated and seamless manner.
+
+#### Story 3
+- As an operator of an OCP Dedicated Managed Platform, I want to give users flexibility to add as many workers nodes as they want or to enable autoscaling on woker nodes so I need to have ability to resize the control plane instances in a declarative, automated and seamless manner to react quickly to cluster growth.
+
+#### Story 4
+- As a SRE, I want to have consumable API primitives in place to resize Control Plane compute resources so I can develop upper level automation tooling atop. E.g Control Plane autoresizing to support a severe growth peak of the number of worker nodes.
+
+#### Story 5
+- As an operator, I want my cluster infrastructure to be self healing so I want faulty nodes to be remediated automatically for safety scenarios.
+
+#### Story 6
+- As a multi cluster operator, I want to have a universal user experience for managing the Control Plane in a declarative manner across any cloud provider, bare metal and any flavour of the product that have in common the topology assumed in this doc.
+
+#### Story 7
+- As a developer, I want to be able to deploy the smallest possible cluster to save costs, i.e one instance Control Plane.
+
+### Implementation Details/Notes/Constraints [optional]
+
+To satisfy the goals, motivation and stories above this propose a new CRD and controller that manages the Control Plane as a single entity which supports the following features:
+
+#### Declarative horizontal scaling
+##### Scale out
+1 - The controller always reconciles towards expected number of replicas. This must be an odd number.
+2 - Fetch all existing control plane Machine resources by ownerRef. Adopt any other machine having a targeted label.
+3 - Compare with expected replicas number. If expected is higher than current then:
+4 - Check all owned machines have a backed ready node.
+5 - Check all Etcd members for all owned machines are healthy via Cluster Etcd Operator status signalling.
+6 - If (NOT all Etcd members are healthy OR NOT all owned machines have a backed ready node) then controller short circuits here, log, update status and requeue. Else:
+7 - Choose a failure domain.
+8 - Create new Machine object with a templated spec. Go to 1.
```

Comment:
I like the idea of using machinesets, but are they compatible with UPI?

Reply:
> like the idea of using machinesets, but are they compatible with UPI?

Any machine API resource is tied to the the API and controllers being operational regardless of how you bootstrap the product UPI/IPI.

Reply:
> I think we should layer this on top of machinesets instead of machine objects, because each machineset object already defines a failure domain.

Regardless of we layer on machines or machineSets we need something on top to decide which failure domain out of a given list to pick for the machine/machineSet. So we are able to scale out evenly. Nothing does that for you today unless we assume the machineSets would be always pre-created by something external i.e the installer.

Reply:
@abhinavdahiya let me think about this a bit.

Reply:
> Nothing does that for you today unless we assume the machineSets would be always pre-created by something external to the controller i.e the installer.

the installer today creates workers and masters across failure domains..

for workers we have a machineset in each failure domain, and we set replicas so that they are evenly distributed.

expecting that each machineset defines a failure domain is probably a good idea imo...